### ISSUE TITLE
CSV reader on .zst and lag/lead: line order is not preserved 

### ISSUE THREAD

**Issue by rpbouman** [duckdb/duckdb#12933]
### What happens?

When using `read_csv()` to read from a .zst-compressedfile and using `lead(line, 1) over ()`, the positions of the rows in the resultset do not match the positions of the corresponding lines in the file, even though the `preserve_insertion_order` is set to `TRUE` (default). (`lag(line, 1) over ()` shows similar behavior)

WIth a plain, uncompressed text file, the order is as expected.
Interestingly, adding `row_number() over ()` seems to remedy the issue.

### To Reproduce

```sql
SELECT  line
,       lead(line, 1) over () as next_line
FROM    read_csv(
          'https://database.lichess.org/standard/lichess_db_standard_rated_2013-01.pgn.zst'
        , columns = {'line': 'VARCHAR'}
        )
LIMIT 20;
```

output:

```
┌──────────────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│         line         │                                                                next_line                                                                 │
│       varchar        │                                                                 varchar                                                                  │
├──────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ [BlackElo "1855"]    │ [WhiteRatingDiff "+13"]                                                                                                                  │
│ [WhiteRatingDiff ".  │ [BlackRatingDiff "-15"]                                                                                                                  │
│ [BlackRatingDiff ".  │ [ECO "C00"]                                                                                                                              │
│ [ECO "C00"]          │ [Opening "Queen's Pawn Game: Franco-Sicilian Defense"]                                                                                   │
│ [Opening "Queen's .  │ [TimeControl "60+0"]                                                                                                                     │
│ [TimeControl "60+0"] │ [Termination "Normal"]                                                                                                                   │
│ [Termination "Norm.  │                                                                                                                                          │
│                      │ 1. e4 c5 2. d4 e6 3. dxc5 Bxc5 4. Nf3 Be7 5. c4 d5 6. cxd5 exd5 7. exd5 Nf6 8. Nc3 O-O 9. Bg5 h6 10. Bxf6 Bxf6 11. Be2 Re8 12. O-O Bg4.  │
│ 1. e4 c5 2. d4 e6 .  │                                                                                                                                          │
│                      │ [Event "Rated Bullet game"]                                                                                                              │
│ [Event "Rated Bull.  │ [Site "https://lichess.org/4op0p4eh"]                                                                                                    │
│ [Site "https://lic.  │ [White "LEGENDARY_ERFAN"]                                                                                                                │
│ [White "LEGENDARY_.  │ [Black "RookieRook"]                                                                                                                     │
│ [Black "RookieRook"] │ [Result "0-1"]                                                                                                                           │
│ [Result "0-1"]       │ [UTCDate "2013.01.01"]                                                                                                                   │
│ [UTCDate "2013.01..  │ [UTCTime "02:16:18"]                                                                                                                     │
│ [UTCTime "02:16:18"] │ [WhiteElo "1395"]                                                                                                                        │
│ [WhiteElo "1395"]    │ [BlackElo "1525"]                                                                                                                        │
│ [BlackElo "1525"]    │ [WhiteRatingDiff "-18"]                                                                                                                  │
│ [WhiteRatingDiff ".  │ [BlackRatingDiff "+8"]                                                                                                                   │
├──────────────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ 20 rows                                                                                                                                               2 columns │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Query demonstrating line order as it appears in file:

```sql
SELECT  line
FROM    read_csv(
          'https://database.lichess.org/standard/lichess_db_standard_rated_2013-01.pgn.zst'
        , columns = {'line': 'VARCHAR'}
        )
LIMIT 20;
```

Output:
```
┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                        line                                                                        │
│                                                                      varchar                                                                       │
├────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ [Event "Rated Classical game"]                                                                                                                     │
│ [Site "https://lichess.org/j1dkb5dw"]                                                                                                              │
│ [White "BFG9k"]                                                                                                                                    │
│ [Black "mamalak"]                                                                                                                                  │
│ [Result "1-0"]                                                                                                                                     │
│ [UTCDate "2012.12.31"]                                                                                                                             │
│ [UTCTime "23:01:03"]                                                                                                                               │
│ [WhiteElo "1639"]                                                                                                                                  │
│ [BlackElo "1403"]                                                                                                                                  │
│ [WhiteRatingDiff "+5"]                                                                                                                             │
│ [BlackRatingDiff "-8"]                                                                                                                             │
│ [ECO "C00"]                                                                                                                                        │
│ [Opening "French Defense: Normal Variation"]                                                                                                       │
│ [TimeControl "600+8"]                                                                                                                              │
│ [Termination "Normal"]                                                                                                                             │
│                                                                                                                                                    │
│ 1. e4 e6 2. d4 b6 3. a3 Bb7 4. Nc3 Nh6 5. Bxh6 gxh6 6. Be2 Qg5 7. Bg4 h5 8. Nf3 Qg6 9. Nh4 Qg5 10. Bxh5 Qxh4 11. Qf3 Kd8 12. Qxf7 Nc6 13. Qe8# 1-0 │
│                                                                                                                                                    │
│ [Event "Rated Classical game"]                                                                                                                     │
│ [Site "https://lichess.org/a9tcp02g"]                                                                                                              │
├────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│                                                                      20 rows                                                                       │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

### OS:

windows 11 enterprise

### DuckDB Version:

v1.0.0 1f98600c2c

### DuckDB Client:

CLI

### Full Name:

Roland Bouman

### Affiliation:

EPAM Systems BV Netherlands

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a nightly build

### Did you include all relevant data sets for reproducing the issue?

Yes

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by rpbouman** [comment]
Here are a few more observations that may be relevant to finding the cause/fixing it:

#1
With a similar `lag(line, 1) over()` expression, the result is different than from `lead(line, 1) over()`, but in both cases the order doesn't match line order

#2
If you open a new session, and first run the query with `lead()`, and then immediately after that, within the same session, the query with `lag()`, the result of the latter does match the line order!

#3 
Adding `row_number() over () `seems to enforce the correct row order.

**Comment by Mytherin** [comment]
Thanks for the report! I have a fix up in #12982. After that PR the results of the queries are as follows:

```sql
D SELECT  line
  FROM    read_csv(
            'lichess.csv.zst'
          , columns = {'line': 'VARCHAR'}
          )
  LIMIT 10;
┌───────────────────────────────────────┐
│                 line                  │
│                varchar                │
├───────────────────────────────────────┤
│ [Site "https://lichess.org/j1dkb5dw"] │
│ [White "BFG9k"]                       │
│ [Black "mamalak"]                     │
│ [Result "1-0"]                        │
│ [UTCDate "2012.12.31"]                │
│ [UTCTime "23:01:03"]                  │
│ [WhiteElo "1639"]                     │
│ [BlackElo "1403"]                     │
│ [WhiteRatingDiff "+5"]                │
│ [BlackRatingDiff "-8"]                │
├───────────────────────────────────────┤
│                10 rows                │
└───────────────────────────────────────┘

D SELECT  line
  ,       lead(line, 1) over () as next_line
  FROM    read_csv(
            'lichess.csv.zst'
          , columns = {'line': 'VARCHAR'}
          )
  LIMIT 10;
┌───────────────────────────────────────┬────────────────────────┐
│                 line                  │       next_line        │
│                varchar                │        varchar         │
├───────────────────────────────────────┼────────────────────────┤
│ [Site "https://lichess.org/j1dkb5dw"] │ [White "BFG9k"]        │
│ [White "BFG9k"]                       │ [Black "mamalak"]      │
│ [Black "mamalak"]                     │ [Result "1-0"]         │
│ [Result "1-0"]                        │ [UTCDate "2012.12.31"] │
│ [UTCDate "2012.12.31"]                │ [UTCTime "23:01:03"]   │
│ [UTCTime "23:01:03"]                  │ [WhiteElo "1639"]      │
│ [WhiteElo "1639"]                     │ [BlackElo "1403"]      │
│ [BlackElo "1403"]                     │ [WhiteRatingDiff "+5"] │
│ [WhiteRatingDiff "+5"]                │ [BlackRatingDiff "-8"] │
│ [BlackRatingDiff "-8"]                │ [ECO "C00"]            │
├───────────────────────────────────────┴────────────────────────┤
│ 10 rows                                              2 columns │
└────────────────────────────────────────────────────────────────┘

D SELECT  line
  ,       lag(line, 1) over () as next_line
  FROM    read_csv(
            'lichess.csv.zst'
          , columns = {'line': 'VARCHAR'}
          )
  LIMIT 10;
┌───────────────────────────────────────┬───────────────────────────────────────┐
│                 line                  │               next_line               │
│                varchar                │                varchar                │
├───────────────────────────────────────┼───────────────────────────────────────┤
│ [Site "https://lichess.org/j1dkb5dw"] │ NULL                                  │
│ [White "BFG9k"]                       │ [Site "https://lichess.org/j1dkb5dw"] │
│ [Black "mamalak"]                     │ [White "BFG9k"]                       │
│ [Result "1-0"]                        │ [Black "mamalak"]                     │
│ [UTCDate "2012.12.31"]                │ [Result "1-0"]                        │
│ [UTCTime "23:01:03"]                  │ [UTCDate "2012.12.31"]                │
│ [WhiteElo "1639"]                     │ [UTCTime "23:01:03"]                  │
│ [BlackElo "1403"]                     │ [WhiteElo "1639"]                     │
│ [WhiteRatingDiff "+5"]                │ [BlackElo "1403"]                     │
│ [BlackRatingDiff "-8"]                │ [WhiteRatingDiff "+5"]                │
├───────────────────────────────────────┴───────────────────────────────────────┤
│ 10 rows                                                             2 columns │
└───────────────────────────────────────────────────────────────────────────────┘

```

**Comment by rpbouman** [comment]
Awesome! Incredibly impressed by how quickly these things are picked up. Happy :)
